packages/flutter/lib/src/widgets/editable_text.dart

```diff
@@ -4829,83 +4914,90 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
-                            key: _editableKey,
-                            startHandleLayerLink: _startHandleLayerLink,
-                            endHandleLayerLink: _endHandleLayerLink,
-                            inlineSpan: buildTextSpan(),
-                            value: _value,
-                            cursorColor: _cursorColor,
-                            backgroundCursorColor: widget.backgroundCursorColor,
-                            showCursor: _cursorVisibilityNotifier,
-                            forceLine: widget.forceLine,
-                            readOnly: widget.readOnly,
-                            hasFocus: _hasFocus,
-                            maxLines: widget.maxLines,
-                            minLines: widget.minLines,
-                            expands: widget.expands,
-                            strutStyle: widget.strutStyle,
-                            selectionColor: _selectionOverlay?.spellCheckToolbarIsVisible ?? false
-                                ? _spellCheckConfiguration.misspelledSelectionColor ?? widget.selectionColor
-                                : widget.selectionColor,
-                            textScaler: effectiveTextScaler,
-                            textAlign: widget.textAlign,
-                            textDirection: _textDirection,
-                            locale: widget.locale,
-                            textHeightBehavior: widget.textHeightBehavior ?? DefaultTextHeightBehavior.maybeOf(context),
-                            textWidthBasis: widget.textWidthBasis,
-                            obscuringCharacter: widget.obscuringCharacter,
-                            obscureText: widget.obscureText,
-                            offset: offset,
-                            rendererIgnoresPointer: widget.rendererIgnoresPointer,
-                            cursorWidth: widget.cursorWidth,
-                            cursorHeight: widget.cursorHeight,
-                            cursorRadius: widget.cursorRadius,
-                            cursorOffset: widget.cursorOffset ?? Offset.zero,
-                            selectionHeightStyle: widget.selectionHeightStyle,
-                            selectionWidthStyle: widget.selectionWidthStyle,
-                            paintCursorAboveText: widget.paintCursorAboveText,
-                            enableInteractiveSelection: widget._userSelectionEnabled,
-                            textSelectionDelegate: this,
-                            devicePixelRatio: _devicePixelRatio,
-                            promptRectRange: _currentPromptRectRange,
-                            promptRectColor: widget.autocorrectionTextRectColor,
-                            clipBehavior: widget.clipBehavior,
+                child: NotificationListener<ScrollNotification>(
+                  onNotification: (ScrollNotification notification) {
+                    _internalScrolling = true;
```

Comment:
Looks like `_internalScrolling` doesn't have to be an instance variable? Making it a parameter would make it slightly easier to understand imo, as the scope of the variable is smaller.

Reply:
Also why do we need to distinguish internal / parent scrolling?

Reply:
I decided to distinguish between internal/parent scrolling because when using only `ScrollNotificationObserver`, `EditableText` received internal scrolling updates too late and the context menu was hidden on `TextSelectionGestureDetector.onTapDown` which prevented the scroll logic from making the toolbar re-appear. `onTapDown` is dispatched even though the `TextSelectionGestureDetector` has not won the arena yet. This is due to the default behavior of `TapAndDragGestureRecognizer`, where it dispatches `onTapDown` after `kPressTimeout = 100ms` if no other recognizer has won the arena. When placing the `NotificationListener` as a direct parent of the internal `Scrollable` we are able to catch the `ScrollNotification` in time to beat the `kPressTimeout`.

Reply:
I was able to get rid of the `_internalScrolling` and the differentiation by just using `ScrollNotificationObserver`, however doing this makes `EditableText` unable to receive any scroll notifications unless it has a parent `ScrollNotificationObserver` (`Scaffold` has one but if someone doesn't use it they won't be able to take advantage of this feature). If we go with this route we should document these details somewhere. A pro of this solution is that we are able to conditionally listen for internal scrolling events (only listen when the toolbar is shown), versus `NotificationListener` where we do not have this control.

Using the previous combination of `NotificationListener` and `ScrollNotificationObserver` allows `EditableText` to still listen to its own internal scrollable notifications without a parent `ScrollNotificationObserver`. To listen to parent scroll events a user would need to place a `ScrollNotificationObserver` somewhere higher in the tree. I think if we go with this solution we need to document this detail somewhere. I can also remove the use of the local member `_internalScrolling` with this solution by checking if the notification context is the same as the internal scrollable context.

Maybe another option would be to place `ScrollNotificationObserver` even higher in the tree in the framework like `MaterialApp` or `WidgetsApp`? This would cover a larger range of users.

Curious to hear thoughts on this @LongCatIsLooong @Piinks

Reply:
Discussed some of this offline, but leaning towards using the combination of NotificationListener and ScrollNotificationObserver so EditableText can get this functionality for it's internal scrollable without having to place a ScrollNotificationObserver as a parent.